Register layout

seromarin opened the pull request:
## Summary
- Completed register layout using reactive forms, validations with validators and custom validators, etc. 
- SCSS migrated from layout training

## Screenshot
![Captura de Pantalla 2020-11-10 a la(s) 5 11 10 p  m](https://user-images.githubusercontent.com/9323975/98739869-204e9280-2378-11eb-99fa-d5fac83a9895.png)
![Captura de Pantalla 2020-11-10 a la(s) 5 11 25 p  m](https://user-images.githubusercontent.com/9323975/98739872-22b0ec80-2378-11eb-95d7-7b23782686d5.png)
![Captura de Pantalla 2020-11-10 a la(s) 5 11 37 p  m](https://user-images.githubusercontent.com/9323975/98739878-247ab000-2378-11eb-996e-e07ea5e4e851.png)
![Captura de Pantalla 2020-11-10 a la(s) 5 11 57 p  m](https://user-images.githubusercontent.com/9323975/98739884-25abdd00-2378-11eb-9b43-b668fda584e3.png)
![Captura de Pantalla 2020-11-10 a la(s) 5 12 09 p  m](https://user-images.githubusercontent.com/9323975/98739889-2775a080-2378-11eb-8c0b-b2e8f908b6cb.png)
![Captura de Pantalla 2020-11-10 a la(s) 5 13 13 p  m](https://user-images.githubusercontent.com/9323975/98739896-293f6400-2378-11eb-9760-c5c0f799625c.png)

## Trello Card
https://trello.com/c/2LA9NmC4/8-register-layout

seromarin committed:
Register layout, SCSS migration, reactive forms & validations

seromarin committed:
Fix password pattern to support at least 1 mayus and 1 number

seromarin committed:
Clean commit

idcuadrowolox commented on wolox-books/angular.json:
```diff
@@ -126,6 +126,10 @@
+  "cli": {
+    "analytics": "ba969946-f146-4c73-ad86-6016808131e4"
+  }
+}
```
add empty line at the end of files

seromarin replied:
Added

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.ts:
```diff
@@ -0,0 +1,35 @@
+    })
+  }
+
+  ngOnInit(): void {
```
remove empty this

seromarin replied:
Removed

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.ts:
```diff
@@ -0,0 +1,35 @@
+export class RegisterComponent implements OnInit {
+
+  registerForm: FormGroup;
+  patternPasswordValidation: RegExp = /.*[A-Z]+.*[0-9]+.*/;
```
add this to constant file

seromarin replied:
New folder to constants and new constant file to pattern password validation

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.spec.ts:
```diff
@@ -0,0 +1,25 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+
```
remove empty line

seromarin replied:
File removed

idcuadrowolox commented on wolox-books/src/app/app.component.scss:
```diff
@@ -0,0 +1,10 @@
+@import './scss/login';
+@import './scss/navbar';
+@import './scss/detail';
+@import './scss/list';
```
put this files in `src/scss` folder and import them in `src/styles.scss`

seromarin replied:
Directory structure and imports fixed

idcuadrowolox commented on wolox-books/src/app/app.component.scss:
```diff
@@ -0,0 +1,10 @@
+@import './scss/login';
+@import './scss/navbar';
+@import './scss/detail';
+@import './scss/list';
```
move component style files to the respective component folder

seromarin replied:
Clean directory structure

idcuadrowolox commented on wolox-books/src/app/scss/_styles.scss:
```diff
@@ -0,0 +1,60 @@
+  font-family: $primary-font;
+  font-size: 20px;
+  height: 20px;
+  margin: 50px 0 110px 0;
```
```suggestion
  margin: 50px 0 110px;
```

seromarin replied:
Fixed

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,32 @@
+    <div class="register__email form-input">
+      <label class="form-input__label" for="email">Email</label>
+      <input class="form-input__input" type="email" name="email" id="email" formControlName="email">
+      <span class="form-input__error" *ngIf="registerForm.controls['email']?.errors?.required && registerForm.controls['email']?.touched">El email es requerido</span>
```
create variable in controller for forms controls and use them here, example:
`this.emailControl = this.registerForm.controls['email'] as FormControl;`

```suggestion
      <span class="form-input__error" *ngIf="emailControl.hasError('required')">
        El email es requerido
      </span>
```

seromarin replied:
Getter created to email form control and changed to hasError() method

seromarin replied:
Getter created to password and password_confirmation too.

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,32 @@
+    </div>
+    <div class="register__password-confirmation form-input">
+      <label class="form-input__label" for="password-confirmation">Confirmación de Password</label>
+      <input class="form-input__input" type="password" name="password-confirmation" id="password-confirmation" formControlName="password-confirmation">
```
when you have long lines it is a good idea to do this to improve readability

```suggestion
      <input
        class="form-input__input"
        type="password"
        name="password-confirmation"
        id="password-confirmation"
        formControlName="password-confirmation">
```

seromarin replied:
I tried...

seromarin committed:
General fixes on directory structure and readability

seromarin committed:
Merge branch 'main' into register-layout

mcallegari10 commented on wolox-books/src/scss/_equalizer.scss:
```diff
@@ -0,0 +1,92 @@
+$equalizer-black: #000;
```
Actually, this file is no longer needed. You can use our own equalizer https://www.npmjs.com/package/wolox-equalizer 😄

seromarin replied:
Sure! Package added and implemented!

mcallegari10 commented on wolox-books/src/app/screens/register/register.component.ts:
```diff
@@ -0,0 +1,44 @@
+  ) {
+    console.log('passwordConfirmation :>> ', passwordConfirmation);
+    this.registerForm = fb.group({
+      'first_name': [ null ],
+      'last_name': [ null ],
```
I think these two fields are required

seromarin replied:
![image](https://user-images.githubusercontent.com/9323975/98860753-230dbe00-2432-11eb-850a-b7a696fb8b1d.png)

seromarin replied:
The card don't mention that the 'first_name' or the 'last_name' are required

mcallegari10 replied:
I think all fields are required for the API to work 🤔

seromarin replied:
OK! I will make all fields required ^^

mcallegari10 commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched && passwordConfirmationFC?.value?.length > 1">Error en la validación del password</span>
+    </div>
+    <button type="submit" class="register-button" [disabled]="!registerForm.valid">Login</button>
```
```suggestion
    <button type="submit" class="register-button" [disabled]="!registerForm.valid">Sign Up</button>
```

seromarin replied:
Fixed

mcallegari10 commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+    </div>
+    <button type="submit" class="register-button" [disabled]="!registerForm.valid">Login</button>
+  </form>
+  <a href="./book-list.html" class="signup-button">Sign Up</a>
```
```suggestion
  <a href="./book-list.html" class="signup-button">Login</a>
```

seromarin replied:
Fixed

mcallegari10 reviewed (changes requested):
Just those details, great job! 🎉

seromarin committed:
General Fixes

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">El email es requerido</span>
+      <span
+        class="form-input__error"
+        *ngIf="emailFC?.errors?.email && emailFC?.touched && emailFC?.value?.length > 5">Formato de email incorrecto</span>
```
```suggestion
        *ngIf="emailFC.hasError('email') && emailFC?.touched">
        Formato de email incorrecto
      </span>
```

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Fix accepted

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Fix incorrect validation

mcallegari10 reviewed (approved):
Check the required fields but I think the card with the API connection will say if those are required or not 😄

seromarin committed:
Refactor app

Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Refactor app

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">La confirmación del password es requerida</span>
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched">Error en la validación del password</span>
```
```suggestion
        *ngIf="passwordConfirmationFC?.hasError('incorrectPassword') && passwordConfirmationFC?.touched">
        Error en la validación del password
      </span>
```

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+        formControlName="password_confirmation">
+      <span
+        class="form-input__error"
+        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">La confirmación del password es requerida</span>
```
```suggestion
        *ngIf="passwordConfirmationFC?.hasError('required') && passwordConfirmationFC?.touched">
        La confirmación del password es requerida
      </span>
```

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">El password es requerido</span>
+      <span
+        class="form-input__error"
+        *ngIf="passwordFC?.hasError('pattern') && passwordFC?.touched">El password debe tener al menos 1 letra mayuscula y 1 numero</span>
```
```suggestion
        *ngIf="passwordFC?.hasError('pattern') && passwordFC?.touched">
        El password debe tener al menos 1 letra mayuscula y 1 numero
      </span>
```

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+      <input class="form-input__input" type="password" name="password" id="password" formControlName="password">
+      <span
+        class="form-input__error"
+        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">El password es requerido</span>
```
```suggestion
        *ngIf="passwordFC?.hasError('required') && passwordFC?.touched">
        El password es requerido
      </span>
```

idcuadrowolox commented on wolox-books/src/app/screens/register/register.component.html:
```diff
@@ -0,0 +1,48 @@
+    <div class="register__email form-input">
+      <label class="form-input__label" for="email">Email</label>
+      <input class="form-input__input" type="email" name="email" id="email" formControlName="email">
+      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">El email es requerido</span>
```
```suggestion
      <span class="form-input__error" *ngIf="emailFC?.hasError('required') && emailFC?.touched">
        El email es requerido
      </span>
```

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Update wolox-books/src/app/screens/register/register.component.html

Co-authored-by: idcuadrowolox <49001977+idcuadrowolox@users.noreply.github.com>

seromarin committed:
Merge branch 'register-layout' of github.com:wolox-training/sr-angular into register-layout
